pkg/operator/configobserver/featuregates/simple_featuregate_reader.go

```diff
@@ -0,0 +1,302 @@
+package featuregates
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"k8s.io/klog/v2"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	v1 "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+type FeatureGateChangeHandlerFunc func(features Features)
+
+type FeatureGateAccess interface {
+	SetChangeHandler(featureGateChangeHandlerFn FeatureGateChangeHandlerFunc)
+
+	Run(ctx context.Context)
+	InitialFeatureGatesObserved() chan struct{}
+	FeatureGatesHaveChangedSinceFirstObserved() chan struct{}
+	CurrentFeatureGates() (enabled []string, disabled []string, err error)
+}
+
+type Features struct {
+	Enabled  []string
+	Disabled []string
+}
+
+type defaultFeatureGateAccess struct {
+	desiredVersion             string
+	missingVersionMarker       string
+	clusterVersionLister       configlistersv1.ClusterVersionLister
+	featureGateLister          configlistersv1.FeatureGateLister
+	featureGateChangeHandlerFn FeatureGateChangeHandlerFunc
+
+	initialFeatureGatesObserved               chan struct{}
+	featureGatesHaveChangedSinceFirstObserved chan struct{}
```

Comment:
Just trying to clarify why we use channels, IIUC these are being used as signals to other threads, the motivation for these being channels over bools is that you don't need a RWMutex around the values if they are channels?

Reply:
> Just trying to clarify why we use channels, IIUC these are being used as signals to other threads, the motivation for these being channels over bools is that you don't need a RWMutex around the values if they are channels?

Yes it is for signals to other threads.  Using channels allows for waiting, which is not as easy with bools.  For example, this is easy to make with channels, hard to make reactive with bools.

```go
select{
case <- featureGateAccessor.InitialFeatureGatesObserved():
	enabled, disabled, _ := featureGateAccessor.CurrentFeatureGates()
	klog.Infof("FeatureGates initialized: enabled=%v  disabled=%v", enabled, disabled)
case <- time.After(1*time.Minute):
	klog.Errorf("timed out waiting for FeatureGate detection")
	return fmt.Errorf("timed out waiting for FeatureGate detection")
}
```